src/samplers/nuts.jl

```diff
@@ -132,47 +111,102 @@ function step(model::Function, spl::Sampler{NUTS}, vi::VarInfo, is_first::Bool)
+                    lj_func::Function, grad_func::Function, stds::Vector; Δ_max=1000) where {T<:Union{Vector,SubArray}}
     doc"""
-      - θ     : model parameter
-      - r     : momentum variable
-      - logu  : slice variable (in log scale)
-      - v     : direction ∈ {-1, 1}
-      - j     : depth of tree
-      - ϵ     : leapfrog step size
-      - H0    : initial H
+      - θ           : model parameter
+      - r           : momentum variable
+      - logu        : slice variable (in log scale)
+      - v           : direction ∈ {-1, 1}
+      - j           : depth of tree
+      - ϵ           : leapfrog step size
+      - H0          : initial H
+      - lj_func     : function for log-joint
+      - grad_func   : function for the gradient of log-joint
     """
     if j == 0
       # Base case - take one leapfrog step in the direction v.
-      θ′, r′, τ_valid = leapfrog(θ, r, 1, v * ϵ, model, vi, spl)
+      θ′, r′, τ_valid = _leapfrog(θ, r, 1, v * ϵ, grad_func)
       # Use old H to save computation
-      H′ = τ_valid == 0 ? Inf : find_H(r′, model, vi, spl)
+      H′ = τ_valid == 0 ? Inf : _find_H(θ′, r′, lj_func, stds)
       n′ = (logu <= -H′) ? 1 : 0
       s′ = (logu < Δ_max + -H′) ? 1 : 0
-      α′ = exp.(min(0, -H′ - (-H0)))
+      α′ = exp(min(0, -H′ - (-H0)))
 
-      θ′, r′, θ′, r′, θ′, getlogp(vi), n′, s′, α′, 1, τ_valid
+      return θ′, r′, θ′, r′, θ′, n′, s′, α′, 1
     else
       # Recursion - build the left and right subtrees.
-      θm, rm, θp, rp, θ′, logp′, n′, s′, α′, n′_α, τ_valid = build_tree(θ, r, logu, v, j - 1, ϵ, H0, model, spl, vi)
+      θm, rm, θp, rp, θ′, n′, s′, α′, n′α = _build_tree(θ, r, logu, v, j - 1, ϵ, H0, lj_func, grad_func, stds)
 
       if s′ == 1
         if v == -1
-          θm, rm, _, _, θ′′, logp′′, n′′, s′′, α′′, n′′_α, τ_valid = build_tree(θm, rm, logu, v, j - 1, ϵ, H0, model, spl, vi)
+          θm, rm, _, _, θ′′, n′′, s′′, α′′, n′′α = _build_tree(θm, rm, logu, v, j - 1, ϵ, H0, lj_func, grad_func, stds)
         else
-          _, _, θp, rp, θ′′, logp′′, n′′, s′′, α′′, n′′_α, τ_valid = build_tree(θp, rp, logu, v, j - 1, ϵ, H0, model, spl, vi)
+          _, _, θp, rp, θ′′, n′′, s′′, α′′, n′′α = _build_tree(θp, rp, logu, v, j - 1, ϵ, H0, lj_func, grad_func, stds)
         end
         if rand() < n′′ / (n′ + n′′)
           θ′ = θ′′
-          logp′ = logp′′
         end
         α′ = α′ + α′′
-        n′_α = n′_α + n′′_α
+        n′α = n′α + n′′α
         s′ = s′′ * (dot(θp - θm, rm) >= 0 ? 1 : 0) * (dot(θp - θm, rp) >= 0 ? 1 : 0)
         n′ = n′ + n′′
       end
 
-      θm, rm, θp, rp, θ′, logp′, n′, s′, α′, n′_α, τ_valid
+      θm, rm, θp, rp, θ′, n′, s′, α′, n′α
+    end
+  end
+
+function _nuts_step(θ, ϵ, lj_func, grad_func, stds)
+
+  d = length(θ)
+  r0 = randn(d)
+  H0 = _find_H(θ, r0, lj_func, stds)
+  logu = log(rand()) + -H0
+
+  θm = θ; θp = θ; rm = r0; rp = r0; j = 0; θ_new = θ; n = 1; s = 1
+  local da_stat
+
+  while s == 1 && j <= 5
```

Comment:
In practise we have to set such a condition for `j` otherwise the sampler can go super slow if the step-size happens to be very small. Stan uses 10 (maximum 2^10 evaluations) as default and 5 (maximum 2^5 evaluations) was originally chosen by Hong because we were bit slow back then. 

Let me make this number an optional argument in the function. We probably need another interface change to support a user specified maximum `j`.